src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -74,7 +74,21 @@ private BoundExpression ReplaceTerminalElementsWithTemps(BoundExpression expr, T
             // Examples:
             // in `expr == (..., ...)` we need to save `expr` because it's not a tuple literal
             // in `(..., expr) == (..., (..., ...))` we need to save `expr` because it is used in a simple comparison
-            return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps);
+            BoundExpression loweredExpr = VisitExpression(expr);
+            if ((object)loweredExpr.Type != null)
+            {
+                BoundExpression value = NullableAlwaysHasValue(loweredExpr);
+                if (value != null)
+                {
+                    // Optimization: if the nullable expression always has a value, we'll replace that value
+                    // with a temp saving that value
+                    BoundExpression savedValue = EvaluateSideEffectingArgumentToTemp(value, initEffects, ref temps);
+                    var objectCreation = (BoundObjectCreationExpression)loweredExpr;
+                    return objectCreation.UpdateArgumentsAndInitializer(ImmutableArray.Create(savedValue), objectCreation.ArgumentRefKindsOpt, objectCreation.InitializerExpressionOpt);
```

Comment:
It is not a contract of `NullableAlwaysHasValue` that it only returns true when the argument is an object creation expression. I expect, for example, that we would expand it to also include certain conversions on top of a creation expression, as is done by the extension method of the same name.

Reply:
Alternately, if that *is* a contract of `NullableAlwaysHasValue*, it should be documented.

---
In reply to: [176492220](https://github.com/dotnet/roslyn/pull/25644#discussion_r176492220) [](ancestors = 176492220)

Reply:
> Alternately, if that is a contract of NullableAlwaysHasValu, it should be documented so we don't break it in the future.

I agree with that, let's add a comment in NullableAlwaysHasValue saing that if it starts returning true for more complicated nodes, the logic here would have to be adjusted accordingly.

---
In reply to: [176493243](https://github.com/dotnet/roslyn/pull/25644#discussion_r176493243) [](ancestors = 176493243,176492220)